src/service.py

```diff
@@ -166,6 +166,20 @@ def restart(self) -> None:
         if not self._startup_check():
             raise ServiceRestartError("Error waiting for the agent service to start")
 
+    def reset_failed_state(self) -> None:
+        """Reset NRestart count of service back to 0.
+
+        The service keeps track of the 'restart-count' and blocks further restarts
+        if the maximum allowed is reached. This count is not reset when the service restarts
+        so we need to do it manually.
+        """
+        try:
+            # Disable protected-access here because reset-failed is not implemented in the lib
+            systemd._systemctl("reset-failed", AGENT_SERVICE_NAME)  # pylint: disable=W0212
+        except systemd.SystemdError:
+            # We only log the exception here as this is not critical
```

Comment:
What are the consequences if it is not reset please?

Reply:
The service will be blocked from restarting again until the count is set back to 0. [By default the max count is 5](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=), [I've set it to 60 in the package](https://github.com/canonical/jenkins-agent-deb/blob/4922cb84c7e951948e83d994b72b80d798b47740/debian/jenkins-agent-1.0.9.jenkins-agent.service#L4).

The thing is the restart count is kept between restarts. 

For example if the connection breaks and the service needed to restart 10 times to reconnect, this count is kept forever if we don't run `systemctl reset-failed`. So after 6 or so incidents like that the service will be blocked from restarting. That's why I think this is needed in `update-status` hook.

Reply:
So if it blocks restarts isn't it a critical issue if the reset fails when the count is such that it would block restarts? Perhaps at that stage we go into error status?

Reply:
That's a good point, my thought process was that during `update-status` if the service is active and running, but the reset fails, the charm should not error since the agent it's still connected, and an error message is sufficient. 

The reset is only ran when the charm is healthy, so I don't think it failing is a big deal. If the reset fails multiple times and the service is blocked from restarting, the charm would error on the next `update-status` which already tells the operator that something is wrong, and hopefully the operator then run `juju debug-log` to see the error messages.

Reply:
I think we can merge this PR and I'll create a follow-up one if needed.

Reply:
Ok I think that is a good approach, thanks!